akka-stream/src/main/scala/akka/stream/impl/fusing/FutureFlow.scala

```diff
@@ -0,0 +1,168 @@
+        override def onPull(): Unit = {
+          hasBeenPulled = true
+        }
+
+        var downstreamCause = OptionVal.none[Throwable]
+        override def onDownstreamFinish(cause: Throwable): Unit = {
+          downstreamCause = OptionVal.Some(cause)
+        }
+
+        def onFuture(futureRes: Try[Flow[In, Out, M]]) = futureRes match {
+          case Failure(exception) =>
+            setKeepGoing(false)
+            innerMatValue.failure(new NeverMaterializedException(exception))
+            failStage(exception)
+          case Success(flow) =>
+            //materialize flow, connect to inlets, feed with potential events and set handlers
+            val initialized = new Initialized(flow)
+            setHandlers(in, out, initialized)
+            setKeepGoing(false)
+        }
+      }
+      class Initialized(flow: Flow[In, Out, M]) extends InHandler with OutHandler {
+        val subSource = new SubSourceOutlet[In](s"${FutureFlow.this}.subIn")
+        val subSink = new SubSinkInlet[Out](s"${FutureFlow.this}.subOut")
+
+        subSource.setHandler {
+          new OutHandler {
+            override def onPull(): Unit = if (!isClosed(in)) tryPull(in)
+
+            override def onDownstreamFinish(cause: Throwable): Unit = {
+              if (!isClosed(in)) {
+                cancel(in, cause)
+              }
+            }
+          }
+        }
+        subSink.setHandler {
+          new InHandler {
+            override def onPush(): Unit = {
+              push(out, subSink.grab())
+            }
+
+            override def onUpstreamFinish(): Unit = {
+              complete(out)
+            }
+
+            override def onUpstreamFailure(ex: Throwable): Unit = {
+              fail(out, ex)
+            }
+          }
+        }
+
+        Try {
+          Source.fromGraph(subSource.source).viaMat(flow)(Keep.right).to(subSink.sink).run()(subFusingMaterializer)
+        } match {
+          case Success(matVal) =>
+            innerMatValue.success(matVal)
+            initializing.upstreamFailure match {
+              case OptionVal.Some(ex) =>
+                subSource.fail(ex)
+              case OptionVal.None =>
+                if (isClosed(in))
+                  subSource.complete()
+            }
+            initializing.downstreamCause match {
+              case OptionVal.Some(cause) =>
+                subSink.cancel(cause)
+              case OptionVal.None =>
+                //todo: should this be invoked before and independently of checking downstreamCause?
+                // in most case if downstream pulls and then closes, the pull is 'lost'. is it possible for some flows to actually care about this? (non-eager broadcast?)
+                if (initializing.hasBeenPulled) {
+                  subSink.pull()
+                }
+            }
+          case Failure(ex) =>
+            innerMatValue.failure(new NeverMaterializedException(ex))
+            failStage(ex)
+        }
```

Comment:
@seglo , @jrudolph ,
so u basically suggest extracting this section to a `connect()` method?
another option is to create the nested ports (subSource, subSink) in the connect method and pass them to the `Initialized`constructor.

is this the direction you guys suggest?

Reply:
I thought @jrudolph was referring to the `Initializing.onFuture` method.  Since this method only references outer stage logic members it should live in the outer stage.  However, I'm not 100% that's what @jrudolph so it's best he respond here.

Is this comment related to https://github.com/akka/akka/pull/28729#discussion_r395285748 ?

Reply:
Yes, everything that's done in `case Success(flow)` in line 122, should be put into a single `def connect` method. There's no real need to create an `Initialized` state then, the handlers could just be anonymous. If you want to give it a name, it could still be a class inside of the `connect` method that closes over the `subSource`, `subSink`. If you want to keep it really explicit, you can also keep the class at top level and pass those in the constructor. Just make sure for clarity, that no initialization code is run in that state class.